File: Laplace/diffinita.c
#include <math.h>
#include <stdio.h>

#define PARTICION 11  // Tamaño de la partición
#define N 9           // Número de incógnitas
#define VISUALIZA 1   // (0) No visualiza la salida, otro valor la visualiza

// Lado derecho de la ecuación diferencial parcial
double LadoDerecho(double x)
{
    double pi = 3.1415926535897932384626433832;
    return -pi * pi * cos(pi * x);
}

// Resuelve Ax=b usando el método Jacobi
void Jacobi(double A[N][N], double x[], double b[], int n, int iter)
{
    int i, j, m;
    double sum;
    double xt[N] = {0};  // Inicializar a cero

    for (m = 0; m < iter; m++) {
        for (i = 0; i < n; i++) {
            sum = 0.0;
            for (j = 0; j < n; j++) {
                if (i == j) continue;
                sum += A[i][j] * x[j];
            }
            if (A[i][i] == 0.0) return;
            xt[i] = (b[i] - sum) / A[i][i];
        }
        for (i = 0; i < n; i++)
            x[i] = xt[i];
    }
}

// Resuelve Ax=b usando el método Gauss-Seidel
void Gauss_Seidel(double A[N][N], double x[], double b[], int n, int iter)
{
    int i, j, m;
    double sum;

    for (m = 0; m < iter; m++) {
        for (i = 0; i < n; i++) {
            sum = 0.0;
            for (j = 0; j < n; j++) {
                if (i == j) continue;
                sum += A[i][j] * x[j];
            }
            if (A[i][i] == 0.0) return;
            x[i] = (b[i] - sum) / A[i][i];
        }
    }
}

int main()
{
    double xi = -1.0;
    double xf = 2.0;
    double vi = -1.0;
    double vf = 1.0;
    int n = PARTICION;
    double h = (xf - xi) / (n - 1);

    double A[N][N] = {0};  // Inicializar toda la matriz a cero
    double b[N] = {0};
    double x[N] = {0};

    double R = 1.0 / (h * h);
    double P = -2.0 / (h * h);
    double Q = 1.0 / (h * h);

    // Primer renglón
    A[0][0] = P;
    A[0][1] = Q;
    b[0] = LadoDerecho(xi) - vi * R;

    // Renglones intermedios
    for (int i = 1; i < N - 1; i++) {
        A[i][i - 1] = R;
        A[i][i] = P;
        A[i][i + 1] = Q;
        b[i] = LadoDerecho(xi + h * i);
    }

    // Renglón final (uso de h*(N-1) y coeficiente R para la frontera)
    A[N - 1][N - 2] = R;
    A[N - 1][N - 1] = P;
    b[N - 1] = LadoDerecho(xi + h * (N - 1)) - vf * R;

    // Resolver
    Gauss_Seidel(A, x, b, N, 1000);
    Jacobi(A, x, b, N, 1000);

    FILE *file = fopen("resultados.dat", "w");
    if (!file) {
        perror("Error al abrir resultados.dat");
        return 1;
    }

    for (int i = 0; i < N; i++) {
        fprintf(file, "%f %f\n", xi + i * h, x[i]);
    }
    fclose(file);

    printf("Resultados escritos en 'resultados.dat'\n");
    return 0;
}

/*
Compilar con:
    gcc programa.c -lm -o programa
*/

File: Laplace/diffinita.f90
! diffinita.f90 modificado para recibir N desde línea de comandos
program fdm1d
    implicit none
    integer :: i, N, nn, ios
    real*8, allocatable :: A(:,:), b(:), x(:)
    real*8 :: xi, xf, vi, vf, h, R, P, Q, y
    character(len=32) :: arg
    
    ! Leer argumento N (nodos interiores)
    call get_command_argument(1, arg, status=ios)
    if (ios /= 0) then
        print *, "Error: debe pasar N (nodos interiores) como argumento."
        stop 1
    end if
    read(arg, *, iostat=ios) N
    if (ios /= 0 .or. N < 1) then
        print *, "Error: N debe ser entero positivo."
        stop 1
    end if
    
    nn = N + 2    ! incluir puntos de frontera
    
    ! Parámetros del problema
    xi = -1.0d0
    xf =  2.0d0
    vi = -1.0d0
    vf =  1.0d0
    h  = (xf - xi) / (nn - 1)
    
    ! Reservar memoria
    allocate(A(N,N), b(N), x(N))
    A = 0.0d0
    b = 0.0d0
    x = 0.0d0
    
    R =  1.0d0 / (h*h)
    P = -2.0d0 / (h*h)
    Q =  1.0d0 / (h*h)
    
    ! Primer renglón
    A(1,1) = P
    if (N > 1) A(1,2) = Q
    call ladoDerecho(xi + h, y)
    b(1) = y - vi*R
    
    ! Renglones intermedios
    do i = 2, N-1
        A(i,i-1) = R
        A(i,i)   = P
        A(i,i+1) = Q
        call ladoDerecho(xi + i*h, y)
        b(i) = y
    end do
    
    ! Último renglón
    if (N > 1) A(N,N-1) = R
    A(N,N) = P
    call ladoDerecho(xi + N*h, y)
    b(N) = y - vf*Q
    
    ! Resolver sistema por Gauss-Seidel
    call gaussSeidel(A, x, b, N, 10000)
    
    ! Guardar resultados
    open(unit=10, file="resultado_"//trim(adjustl(arg))//".dat", status="replace")
    write(10, '(F10.5,1X,F10.5)') xi, vi
    do i = 1, N
        write(10, '(F10.5,1X,F10.5)') xi + i*h, x(i)
    end do
    write(10, '(F10.5,1X,F10.5)') xf, vf
    close(10)
    
end program fdm1d

subroutine ladoDerecho(x, y)
    implicit none
    real*8, intent(in)  :: x
    real*8, intent(out) :: y
    real*8, parameter :: pi = 3.14159265358979323846d0
    y = -pi*pi * cos(pi*x)
end subroutine ladoDerecho

subroutine gaussSeidel(a, x, b, nx, iter)
    implicit none
    integer, intent(in)    :: nx, iter
    real*8, intent(in)     :: a(nx,nx), b(nx)
    real*8, intent(inout)  :: x(nx)
    integer :: i, j, m
    real*8 :: sum
    
    do m = 1, iter
        do i = 1, nx
            sum = 0.0d0
            do j = 1, nx
                if (i /= j) sum = sum + a(j,i)*x(j)
            end do
            x(i) = (b(i) - sum)/a(i,i)
        end do
    end do
end subroutine gaussSeidel